filebeat/input/log/log.go

```diff
@@ -100,10 +102,42 @@ func (f *Log) errorChecks(err error) error {
 		return err
 	}
 
+	// End of file reached
 	if err == io.EOF && f.config.CloseEOF {
 		return err
 	}
 
+	return f.closeChecks()
+}
+
+func (f *Log) wait() {
+	// Wait before trying to read file again. File reached EOF.
+	select {
+	case <-f.done:
+		return
+	case <-time.After(f.backoff):
+	}
+
+	// Increment backoff up to maxBackoff
+	if f.backoff < f.config.MaxBackoff {
+		f.backoff = f.backoff * time.Duration(f.config.BackoffFactor)
+		if f.backoff > f.config.MaxBackoff {
+			f.backoff = f.config.MaxBackoff
+		}
+	}
+}
+
+// Close closes the done channel but no th the file handler
+func (f *Log) Close() {
+	close(f.done)
+	// Note: File reader is not closed here because that leads to race conditions
+}
+
+func (f *Log) closeChecks() error {
+
+	f.closeMutex.Lock()
+	defer f.closeMutex.Unlock()
```

Comment:
Still smells like race conditions. File processing state like `f.offset` and `f.lastTimeRead` is accessed, but potentially modified by another thread.